### ISSUE TITLE
[Proposal] Use the testing framework assertion library for better integration

### ISSUE THREAD

**Issue by mike1808** [jfairbank/redux-saga-test-plan#262]
The current behavior is unmet expectation is to throw an error with a formatted string of unmet expectations. 

For small examples it is not an issue, however, when you have nested objects to assert (e.g. 2 level deep objects) or many effects to except the error message becomes very unreadable. It is hard to find what exactly didn't match.

If we use the testing framework, for example `jest`, assertion library (`expect`) errors will be printed using the library's reported. 

In my case, they will formated by WebStorm reported which have a nice diff view which I missing when using `redux-saga-test-plan`.

I will hack the library and write `expect` instead of current `throw new SagaTestError` to see how it works.

What do you think about this proposal?

**Comment by jp928** [comment]
@mike1808 Thanks for the proposal and pull request would be more than welcomed. Please let me know if you need any assistance.

**Comment by fstoerkle** [comment]
@mike1808 @jp928 here is a [work-in-progress branch for always using `jest-diff`](https://github.com/fstoerkle/redux-saga-test-plan/commit/89e38b9293bdf805c029db8662b767b579584611), but using the `expect` from the test framework makes would be even nicer!

**Comment by borrascador** [comment]
@fstoerkle Where can I find this `expect` method you mentioned? I'm interested in trying to make that change and open a PR

**Comment by mike1808** [comment]
@borrascador it's the built-in `expect` in Jest. Try to use it in https://github.com/jfairbank/redux-saga-test-plan/blob/master/src/expectSaga/expectations.js file instead of `throw`

**Comment by ivan-aksamentov** [comment]
Can't wait this feature!

This would resolve the confusing assertions like:

```
SagaTestError: 
    Assertion 3 failed: put effects do not match

    Expected
    --------
    { '@@redux-saga/IO': true,
      combinator: false,
      type: 'PUT',
      payload: 
       { channel: undefined,
         action: 
          { type: 'FSA_TEST/REPLY_FAILED',
            payload: { params: [Object], error: [Object] },
            error: true } } }

    Actual
    ------
    { '@@redux-saga/IO': true,
      combinator: false,
      type: 'PUT',
      payload: 
       { channel: undefined,
         action: 
          { type: 'FSA_TEST/REPLY_FAILED',
            payload: { params: [Object], error: [Object] },
            error: true } } }
```

In this case `payload.action.payload.error.message` don't match, but you cannot figure it out because the object serialization does not reach the necessary depth.

**Comment by mike1808** [comment]
Hi @jp928 . I have started working on this feature and have the following steps to complete and wanted to discuss them with you:

- [ ] Change effect serialization to make them easier to read (see screenshots)
- [ ] Change expectations implementation to return the matching result and in index file of redux-saga-test-plan detect the testing engine and bind those matching results to the test runner. I bound only to jest. (Inspired by [enzyme-matchers](https://github.com/FormidableLabs/enzyme-matchers) where they have generic API and binding for different test runners/expectations libraries)

Results so far:

In terminal when checking for `PUT` actions.

![image](https://user-images.githubusercontent.com/3942554/58915109-1b8ea500-86d5-11e9-9e54-0156bc49a737.png)

In WebStorm for the same test

![image](https://user-images.githubusercontent.com/3942554/58915139-3234fc00-86d5-11e9-9e50-9bb13ebadd17.png)

With supported diff view

![image](https://user-images.githubusercontent.com/3942554/58915164-40831800-86d5-11e9-9b2e-9e4ce3b330e3.png)

**Comment by ivan-aksamentov** [comment]
@mike1808 Looks great! Is it possible to also add diffs for partial matches?

**Comment by mike1808** [comment]
@ivan-aksamentov what do you mean by partial match? Can you describe a use case? If that something that possible in Jest it'd be possible here because custom watchers have access to Jest API.

**Comment by mike1808** [comment]
@ivan-aksamentov if you referring to your previous message, I think, that it's not so trivial, as redux-test-saga-plan doesn't know which particular effect you have expected (you may have the same action dispatched with different payloads). However, I think, the library can suggest you have a look at particular actions which were dispatched by comparing `type` then other values in `payload`.

**Comment by ivan-aksamentov** [comment]
@mike1808 Like in the "With supported diff view", but in console
https://pbs.twimg.com/media/D4UMeMUUwAIAfGD.png
I think [`jest-diff`](https://www.npmjs.com/package/jest-diff) is what implements this feature in jest.

**Comment by mike1808** [comment]
@ivan-aksamentov yes, currently I'm focused on integration tests. For unit testing that is 100% doable. Jest exposes deep comparison which will produce the same output as expect().isEqual().

I'll add unit testing support to my to-do list, thanks

**Comment by dbartholomae** [comment]
Any news here? Is there some way I could help? :)

**Comment by mike1808** [comment]
@dbartholomae Ah, I joined a new company and started doing Go and kinda forgot about this PR. I'll try to bring to some workable state and commit this weekend. So everyone who's interested can continue the work.

**Comment by dbartholomae** [comment]
@mike1808 Thanks! If you don't find time to bring it in a workalbe state, please feel free to commit push as-is :)

**Comment by mike1808** [comment]
For those who are interested. Unfortunately, I didn't have time to make the code more organized, however, Jest integration works perfectly. I tested my previous project which had a lot of saga tests using this library. 

The code is available in my forked repo on this branch https://github.com/mike1808/redux-saga-test-plan/tree/beta-jest

You can try on your projects by installing `jest-redux-saga-test-plan@4.0.0-beta.5` 

Few caveats:

1. This branch is created from `upgrade-to-babel7` as its new versions of babel was required to use the new Jest version.
2. The compatibility for old Node.js might be broken because I upgraded Babel and removed some polyfills in the code (like `Map`), so my fork might not work on Node.js < 8
3. The main code changes are in src/shared/serializeEffect.js file. There you can find that some effects like CANCELLED, FLUSH, CANCEL, JOIN are not implemented.
4. The changes are only for `expectSaga`. I didn't touch the unit testing API.

If you have any question about code, feel free to write here, or, if you want to continue my work, we can set up a zoom/skype call and I'll share all context I have.

cc @dbartholomae @vov462 @Taguhi

**Comment by jp928** [comment]
Mark this need to be merged before a stable release 4.0.0
